filebeat/inputsource/common/handler.go

```diff
@@ -0,0 +1,95 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package common
+
+import (
+	"bufio"
+	"net"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/filebeat/inputsource"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+// HandlerFactory returns a ConnectionHandler func
+type HandlerFactory func(config ListenerConfig) ConnectionHandler
+
+// ConnectionHandler interface provides mechanisms for handling of incoming connections
+type ConnectionHandler func(CloseRef, net.Conn) error
+
+// MetadataFunc defines callback executed when a line is read from the split handler.
+type MetadataFunc func(net.Conn) inputsource.NetworkMetadata
+
+// SplitHandlerFactory allows creation of a handler that has splitting capabilities.
+func SplitHandlerFactory(family Family, metadataCallback MetadataFunc, callback inputsource.NetworkFunc, splitFunc bufio.SplitFunc) HandlerFactory {
+	return func(config ListenerConfig) ConnectionHandler {
+		return ConnectionHandler(func(closer CloseRef, conn net.Conn) error {
+			metadata := metadataCallback(conn)
+			maxMessageSize := uint64(config.MaxMessageSize)
+
+			var log *logp.Logger
+			if family == FamilyUnix {
+				// unix sockets have an empty `RemoteAddr` value, so no need to capture it
+				log = logp.NewLogger("split_client")
```

Comment:
Sometimes it's nice to pass in a parent logger and create a child from it. The child logger would retain the context from its parent like the `path` in the unix case and then the logger name gets appended like `input.unix.split_func`. But then you need a new to add another function parameter to an already big signature.